src/main/java/redis/clients/jedis/JedisClusterCommand.java

```diff
@@ -79,61 +101,114 @@ public T runWithAnyNode() {
-        //We need this because if node is not reachable anymore - we need to finally initiate slots
-        //renewing, or we can stuck with cluster state without one node in opposite case.
-        //But now if maxAttempts = [1 or 2] we will do it too often.
-        //TODO make tracking of successful/unsuccessful operations for node - do renewing only
-        //if there were no successful responses from this node last few seconds
+  /**
+   * Related values should be reset if <code>TRUE</code> is returned.
+   *
+   * @param attemptsLeft
+   * @param consecutiveConnectionFailures
+   * @param doneDeadline
+   * @return true - if some actions are taken
+   * <br /> false - if no actions are taken
+   */
+  private boolean handleConnectionProblem(int attemptsLeft, int consecutiveConnectionFailures, Instant doneDeadline) {
+    if (this.maxAttempts < 3) {
+      // Since we only renew the slots cache after two consecutive connection
+      // failures (see consecutiveConnectionFailures above), we need to special
+      // case the situation where we max out after two or fewer attempts.
+      //
+      // Otherwise, on two or fewer max attempts, the slots cache would never be
+      // renewed.
+      if (attemptsLeft == 0) {
         this.connectionHandler.renewSlotCache();
+        return true;
       }
+      return false;
+    }
 
-      return runWithRetries(slot, attempts - 1, tryRandomNode, redirect);
-    } catch (JedisRedirectionException jre) {
-      // if MOVED redirection occurred,
-      if (jre instanceof JedisMovedDataException) {
-        // it rebuilds cluster's slot cache recommended by Redis cluster specification
-        this.connectionHandler.renewSlotCache(connection);
-      }
+    if (consecutiveConnectionFailures < 2) {
+      return false;
+    }
 
-      // release current connection before recursion
-      releaseConnection(connection);
-      connection = null;
+    sleep(getBackoffSleepMillis(attemptsLeft, doneDeadline));
+    //We need this because if node is not reachable anymore - we need to finally initiate slots
+    //renewing, or we can stuck with cluster state without one node in opposite case.
+    //TODO make tracking of successful/unsuccessful operations for node - do renewing only
+    //if there were no successful responses from this node last few seconds
+    this.connectionHandler.renewSlotCache();
+    return true;
+  }
 
-      return runWithRetries(slot, attempts - 1, false, jre);
-    } finally {
-      releaseConnection(connection);
+  private static long getBackoffSleepMillis(int attemptsLeft, Instant deadline) {
+    if (attemptsLeft <= 0) {
+      return 0;
+    }
+
+    long millisLeft = Duration.between(Instant.now(), deadline).toMillis();
+    if (millisLeft < 0) {
+      // TODO: change to JedisClusterOperationException or a new sub-class of it
+      throw new JedisClusterMaxAttemptsException("Deadline exceeded");
+    }
+
+    return millisLeft / (attemptsLeft * attemptsLeft);
```

Comment:
Why divide by the square of attempts? (Just wondering if it is specific or an arbitrary decision?)

Reply:
Kind of arbitrary. Not squared would work as well. The important thing is that the sleeps get longer and longer the fewer iterations that you have left, and that the last sleep uses up all of the time remaining.

**Example**: Let's say you have max 10s and max 5 attempts.

At time 0.0, 5 attempts left, sleep `(10s/(5*5))` = **0.4s**
At time 0.4, 4 attempts left, sleep `(9.6s/(4*4))` = **0.6s**
At time 1.0, 3 attempts left, sleep `(9s/(3*3))` = **1.0s**
At time 2.0, 2 attempts left, sleep `(8/(2*2))` = **2.0s**
At time 6.0, 1 attempts left, sleep `(4/(1*1))` = **4.0s**